---
title: "Linear modeling of RNA-seq data with Limma-Voom: Ancestry Effects"
output:
  html_document:
    df_print: paged
---
Author: Tyler Hansen
Date: 6.6.2024

I previously modeled the infection effect at each timepoint to understand how gene expression changes over time in response to Mtb infection. Now I want to identify how European/African ancestry impacts this response. To do this, I plan to do two different types of analysis. 

1. Population-specific DEs at each timepoint using all expressed genes as input. 
  -Nested model: ~ 0 + Infection + Timepoint + Infection:Timepoint + (African_admixture:Infection):Timepoint + technical confounders[RNA_concentration + perc_Dups + flow_cell + M_Aligned]

2. Population-specific DRs at each timepoint using all expressed genes as input. 
  -Nested model for each timepoint: ~ 0 + Timepoint + (African_admixture*Infection):Timepoint + Donor_ID +  technical confounders[RNA_concentration + perc_Dups + flow_cell + M_Aligned]
  
* For both 1 and 2, the follow-up analyses will be the following: 
  - Line/point plot of number of popDEs and popDRs over time. 
  - Possibly GSEA over time like I did for infection effects
  - Distribution of pathway score over time for the two ancestries (katie paper 2c and 2dd)
  - Overlap of infection effect genes and ancestry genes (venn diagrams).
  - Cluster ancestry effects and perform reactome enrichment analysis. 
  
Analyses done here will likely make up figure 3 (popDEs) and figure 4 (popDRs). 

## Setup

Packages
```{r}
suppressPackageStartupMessages(library(tidyverse))
library(limma)
library(edgeR)
```

Three tables
```{r}
#read in GTF
gtf <- rtracklayer::readGFFAsGRanges("data/Homo_sapiens.GRCh38.87.gtf")

#read in SI
sample_info <- readRDS("data/metadata_batch1-only_approach1_all-samples.rds")

#read in cts
cts <- readRDS("data/cts_batch1-only_approach1_all-samples.rds")
```

## Model Ancestry:

Set up design
```{r}
design <- model.matrix(~ 0 + (Infection + African_admixture:Infection):Timepoint + flow_cell + perc_GC + perc_Aligned + perc_Dups, data = sample_info)
```

Voom transform counts
```{r}
dge <- DGEList(cts)
dge <- calcNormFactors(dge)
v <- voom(dge, design, plot = T)
```

Fit and calc stats
```{r}
vfit <- lmFit(v, design)
vfit <- eBayes(vfit)
```

Check model is behaving as expected:
```{r, fig.width= 10, fig.height=6}
res <- topTable(vfit, coef = "InfectionMtb:African_admixture:TimepointT48", number = 6)
top_v <- v$E[row.names(res), ] %>% as.data.frame() %>% rownames_to_column(var = "gene_id") %>% 
  pivot_longer(cols = !gene_id, names_to = "sample", values_to = "voom_cts") %>% 
  left_join(rownames_to_column(sample_info, var = "sample"), by = "sample") %>% 
  mutate(symbol = str_split_i(gene_id, pattern = "_", i = 2)) %>% dplyr::filter(Timepoint == "T48")

ggplot(top_v, aes(x = Infection, y = voom_cts, color = self_reported_ethnicity, fill = self_reported_ethnicity)) +
  geom_boxplot(alpha = 0.5, color = "black") +
  facet_wrap(~symbol, ncol = 3, scales = "free") +
  scale_fill_manual(values = c("darkorange", "darkblue")) +
  scale_color_manual(values = c("darkorange", "darkblue")) +
  ggtitle("T48") +
  theme_bw()
```

```{r, fig.width= 10, fig.height=6}
res <- topTable(vfit, coef = "InfectionNI:African_admixture:TimepointT48", number = 6)
top_v <- v$E[row.names(res), ] %>% as.data.frame() %>% rownames_to_column(var = "gene_id") %>% 
  pivot_longer(cols = !gene_id, names_to = "sample", values_to = "voom_cts") %>% 
  left_join(rownames_to_column(sample_info, var = "sample"), by = "sample") %>% 
  mutate(symbol = str_split_i(gene_id, pattern = "_", i = 2)) %>% dplyr::filter(Timepoint == "T48")

ggplot(top_v, aes(x = Infection, y = voom_cts, color = self_reported_ethnicity, fill = self_reported_ethnicity)) +
  geom_boxplot(alpha = 0.5, color = "black") +
  facet_wrap(~symbol, ncol = 3, scales = "free") +
  scale_fill_manual(values = c("darkorange", "darkblue")) +
  scale_color_manual(values = c("darkorange", "darkblue")) +
  ggtitle("T48") +
  theme_bw()
```

This looks pretty good. Now lets calculate qvalue. 

### Compute emperical p values and apply the story fdr correction to these values
The q value adjustment for multiple testing correction seems to be the superior method to FWER and Benjamini-Hochberg. The only issue with it is when the pvalue distribution is not uniform. In most cases, this is fine, but it is best to use an empirical approach where the pvalues are first corrected based on null pvalues obtained from permutations. 

To do this, we have the following steps:
1. Permute the variable of interest and calculate the null pvalues, for at least 10 iterations.
2. Adjust the observed pvalues based on the null/expected pvalues using the emPvals() function from the qvalue package. 
3. Use the adjusted p values to calculate storey q values. Use these to threshold genes. 

This is the same approach used for the infection effects. 

#### Permute the variable of interest and calculate the null pvalues, for at least 10 iterations. 
Permute the the infection column within timepoint so that infection is random and the null is the random effect of that. 
```{r}
#function to permute the variable of interest
permute_si_variable_within_category <- function(var, si, cat, n = 10) {
  # var is the variable of interest
  # si is the sample info/metadata
  # cat is the category/variable to permute within (must be a factor)
  # n is the number of iterations
  # this function returns a list of permuted sample info data.frames 
  
  # create dummy list
  list_x <- list()
  
  #split the si into the individual factors of the cat and permute var
  for (category in levels(si[,cat])) {
    #split si
    x <- si[si[,cat] == category, ]
    
    #permute
    x <- rsample::permutations(x, permute = all_of(var), times = n)
    
    #add to list
    list_x[[category]] <- x
  }

  #for each iteration, extract permutations for each timepoint and join back together
  
  #make another dummy list
  list_y <- list()
  
  for (i in seq(1, n)) {
    #make dummy df 
    df <- data.frame()
    
    for (category in levels(si[,cat])) {
        x <- list_x[[category]] 
        x <- x$splits[[i]] %>% rsample::analysis() %>% as.data.frame()
        df <- rbind(df, x)
    }
    
    #add df to list
    list_y[[i]] <- df
  }
  #return the list
  return(list_y)
}
```

```{r}
set.seed(1993)
#execute function:
permuted_si <- permute_si_variable_within_category(var = "African_admixture", si = sample_info, cat = "Timepoint", n = 10)
```

Fit permuted data to linear models, use the same model as done in the observed analysis
```{r}
fit_permutations <- function(perms, formula, v, n =  10) {
  # perms is the permuted sample info
  # formula is the formula used in the respective analysis provided in quotes (ex: "~ Timepoint")
  # v are the voom tramsformed counts to input into the random/null models. 
  # n is the number of iterations performed above. 
  # Returns a list of model fits for the permuted data
  
  # create dummy list
  list_x <- list()
  
  #extract each iteration and set_up random models:
    for (i in seq(1, n)) {
      #set up model for ith iteration
      m_rand <- model.matrix(as.formula(formula), data = as.data.frame(perms[i]))
    
      #fit model for each iteration
      vfit_rand <- lmFit(v, m_rand)
      
      #compute stats
      vfit_rand <- eBayes(vfit_rand)
      
      #add to list
      list_x <- rlist::list.append(list_x, vfit_rand)
    }
  return(list_x)
}
```

```{r}
permuted_fits <- fit_permutations(perms = permuted_si, formula = "~ 0 + (Infection + African_admixture:Infection):Timepoint + flow_cell + perc_GC + perc_Aligned + perc_Dups", v = v, n = 10)
```

Calculate null pvalues
```{r}
get_null_pvals <- function(perm_fits, n =  10) {
  # perm_fits is the list of model fits for permuted data
  # n is the number of iterations performed above. 
  # cat is the category/variable that was permuted within (must be a factor). This is used to identify the individual categories to get pvalues for. 
  # Returns a vector of pvalues for the contrast between Mtb and NI
  
  # create dummy df
  df <- data.frame()
  
  #for each iteration, extract the null pvalues between NI and Mtb:
    for (i in seq(1, n)) {
      #for each iteration, append p.vals to the dummy df
      x <- as.data.frame(perm_fits[[i]]$p.value)
      df <- dplyr::bind_rows(df, x)
    }
  return(df)
}
```

```{r}
null_pvalues <- get_null_pvals(perm_fits = permuted_fits, n = 10)
```

Check the distribution of null pvalues
```{r}
hist(as.matrix(null_pvalues)[, 45:82])
hist(vfit$p.value[, 45:82])
```


For each nested model, calculate p-values from a set of observed test statistics and simulated null test statistics. 

Note: the output of empPval needs to be corrected by 1-value. Tauras explained this is because the function expects effect sizes by default, so it returns the proportion which are larger. If I were to plot empP vs. P, I'd find they're inversely correlated. 

In both models TimepointT1:African_admixture couldn't be measured - probably due to low variance. So skip it by removing from the timepoint_list setup. 

```{r}
#make pval and se dfs
obs_pvals <- as.data.frame(vfit$p.value)
SE <- sqrt(vfit$s2.post) * vfit$stdev.unscaled

#setup list to iterate through
timepoint_list <- list()
for (t in c("T1", "T2","T3","T4","T5","T6","T7","T8","T9","T10","T12","T14","T16","T18","T24","T30","T36","T42","T48")) {
  time <- paste0("InfectionNI:African_admixture:Timepoint", t)
  timepoint_list <- append(timepoint_list, time)
  time <- paste0("InfectionMtb:African_admixture:Timepoint", t)
  timepoint_list <- append(timepoint_list, time)
}

#run through list to extract qvalues for each
for (i in timepoint_list) {
  #get qvals
  emperical_pvals <- qvalue::empPvals(stat = -log10(obs_pvals[, i]), stat0 = -log10(null_pvalues[, i]), pool = T)
  qvals <- qvalue::qvalue(p = emperical_pvals)
  
  #merge with observed pvalues
  tmp1 <- data.frame(obs_pval = obs_pvals[, i], empirical_qval = qvals$qvalues)
  
  #get pvalues just for the timepoint
  tmp2 <- dplyr::select(obs_pvals, all_of(i))
  colnames(tmp2) <- "p.value"
  
  #join to each gene
  tmp3 <- bind_cols(tmp2, tmp1) %>% rownames_to_column("gene_id")
  
  #stop if the row order is not the same
  if (all(tmp3$p.value == tmp3$obs_pval) == FALSE) {
    stop()
  }
  
  #remove redundant col
  tmp3 <- dplyr::select(tmp3, -p.value)
  
  #get the stats
  tmp4 <- data.frame(coefficients = c(vfit$coefficients[, i]), 
                     t_statistic = c(vfit$t[, i]), 
                     coef_se = c(SE[, i]), 
                     p.value = c(vfit$p.value[, i])) %>% rownames_to_column("gene_id")
  
  #join
  tmp5 <- inner_join(tmp4, tmp3, by = "gene_id")
  
  #stop if join doesn't work properly
  if (all(tmp5$p.value == tmp5$obs_pval) == FALSE) {
    stop()
  }
  #remove redundant col 
  tmp5 <- dplyr::select(tmp5, -obs_pval)

  #assign tmp5
  assign(value = tmp5, x = paste0(i))
  
  #remove tmps
  rm(tmp1, tmp2, tmp3, tmp4, tmp5)
}
```

### Number of DEGs per timepoint
Now that we have emprical qvalues, we will threshold with qval.

Report number of DEGs via qval

```{r}
#make dummy df
popDE_count <- data.frame(dummy = c(1))

#loop through timepoints and add the number of genes passing the empirical qvalue to the dummy data frame
for (i in timepoint_list){
  popDEs <- dplyr::filter(get(i), empirical_qval < 0.05) %>% nrow() %>% as.numeric()
  popDE_count[[i]] <- popDEs
}

#remove dummy column, transpose axis, and add rownames to column
popDE_count <- dplyr::select(popDE_count, -dummy) %>% as.matrix() %>% t() %>% as.data.frame() %>% rownames_to_column("Sample") 

#extract timepoint and rename V1
popDE_count <- mutate(popDE_count, Timepoint = str_extract(Sample, "(?<=Timepoint)T\\d+")) %>% mutate(Infection = str_extract(Sample, "(?<=Infection)[^:]+")) %>% rename("popDE_count" = V1)

#relevel timepoint factor
popDE_count$Timepoint <- factor(popDE_count$Timepoint, levels = c("T1","T2","T3","T4","T5","T6","T7","T8","T9","T10","T12","T14","T16","T18","T24","T30","T36","T42","T48"))
```

Plot
```{r}
ggplot(popDE_count, aes(x = Timepoint, y = popDE_count, color = Infection, group = Infection)) +
  geom_line() +
  geom_text(aes(label = popDE_count), vjust = -0.5) + 
  geom_point(size = 3) +
  scale_color_manual(values = c("red", "grey60")) +
  labs(y = "Number of popDEs", x  ="Timepoint",
       title = "Number of popDEs per timepoint", 
       subtitle = "Thresholds: qval < 0.05") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"), plot.subtitle = element_text(hjust = 0.5, face = "italic")) +
  coord_cartesian(ylim = c(0, 8000))
```


## PopDRs

Model!
```{r}
#Set designs for each model. Ancestry is condition-specific. 
design_popDR <- model.matrix(~ 0 + Timepoint:African_admixture*Infection + Timepoint:Infection + Donor_ID + Timepoint + flow_cell + perc_GC + perc_Dups + perc_Aligned, data = sample_info)

#Voom transform 
v_popDR <- DGEList(cts) %>% calcNormFactors() %>% voom(design_popDR)

#Fit to models
vfit_popDR <- lmFit(v_popDR, design_popDR) %>% eBayes()
```

```{r, fig.width= 10, fig.height=6}
res <- topTable(vfit_popDR, coef = "TimepointT48:African_admixture:InfectionMtb", number = 24)
top_v <- v_popDR$E[row.names(res), ] %>% as.data.frame() %>% rownames_to_column(var = "gene_id") %>% 
  pivot_longer(cols = !gene_id, names_to = "sample", values_to = "voom_cts") %>% 
  left_join(rownames_to_column(sample_info, var = "sample"), by = "sample") %>% 
  mutate(symbol = str_split_i(gene_id, pattern = "_", i = 2)) %>% dplyr::filter(Timepoint == "T48")

ggplot(top_v, aes(x = Infection, y = voom_cts, color = self_reported_ethnicity, fill = self_reported_ethnicity)) +
  geom_boxplot(alpha = 0.5, color = "black") +
  facet_wrap(~symbol, ncol = 3, scales = "free") +
  scale_fill_manual(values = c("darkorange", "darkblue")) +
  scale_color_manual(values = c("darkorange", "darkblue")) +
  ggtitle("T48") +
  theme_bw()
```

### Compute emperical p values and apply the story fdr correction to these values 

Use the previous permited si, since this is exactly how we want to do it. 
```{r}
permuted_fits <- fit_permutations(perms = permuted_si, formula = "~ 0 + Timepoint:African_admixture*Infection + Timepoint:Infection + Donor_ID + Timepoint + flow_cell + perc_GC + perc_Dups + perc_Aligned", v = v, n = 10)
```

Calculate null pvalues
```{r}
null_pvalues <- get_null_pvals(perm_fits = permuted_fits, n = 10)
```

Check the distribution of null pvalues
```{r}
hist(as.matrix(null_pvalues)[, 75:93])
hist(vfit_popDR$p.value[, 75:93])
```

For each nested model, calculate p-values from a set of observed test statistics and simulated null test statistics. 

```{r}
#make pval and se dfs
obs_pvals <- as.data.frame(vfit_popDR$p.value)
SE <- sqrt(vfit_popDR$s2.post) * vfit_popDR$stdev.unscaled

#setup list to iterate through
timepoint_list <- list()
for (t in c("T1", "T2","T3","T4","T5","T6","T7","T8","T9","T10","T12","T14","T16","T18","T24","T30","T36","T42","T48")) {
  time <- paste0("Timepoint", t, ":African_admixture:InfectionMtb")
  timepoint_list <- append(timepoint_list, time)
}

#run through list to extract qvalues for each
for (i in timepoint_list) {
  #get qvals
  emperical_pvals <- qvalue::empPvals(stat = -log10(obs_pvals[, i]), stat0 = -log10(null_pvalues[, i]), pool = T)
  qvals <- qvalue::qvalue(p = emperical_pvals)
  
  #merge with observed pvalues
  tmp1 <- data.frame(obs_pval = obs_pvals[, i], empirical_qval = qvals$qvalues)
  
  #get pvalues just for the timepoint
  tmp2 <- dplyr::select(obs_pvals, all_of(i))
  colnames(tmp2) <- "p.value"
  
  #join to each gene
  tmp3 <- bind_cols(tmp2, tmp1) %>% rownames_to_column("gene_id")
  
  #stop if the row order is not the same
  if (all(tmp3$p.value == tmp3$obs_pval) == FALSE) {
    stop()
  }
  
  #remove redundant col
  tmp3 <- dplyr::select(tmp3, -p.value)
  
  #get the stats
  tmp4 <- data.frame(coefficients = c(vfit_popDR$coefficients[, i]), 
                     t_statistic = c(vfit_popDR$t[, i]), 
                     coef_se = c(SE[, i]), 
                     p.value = c(vfit_popDR$p.value[, i])) %>% rownames_to_column("gene_id")
  
  #join
  tmp5 <- inner_join(tmp4, tmp3, by = "gene_id")
  
  #stop if join doesn't work properly
  if (all(tmp5$p.value == tmp5$obs_pval) == FALSE) {
    stop()
  }
  #remove redundant col 
  tmp5 <- dplyr::select(tmp5, -obs_pval)

  #assign tmp5
  assign(value = tmp5, x = paste0(i))
  
  #remove tmps
  rm(tmp1, tmp2, tmp3, tmp4, tmp5)
}
```

### Number of popRDs per timepoint
Now that we have emprical qvalues, we will threshold with qval.

Report number of popDRs via qval

```{r}
#make dummy df
popDR_count <- data.frame(dummy = c(1))

#loop through timepoints and add the number of genes passing the empirical qvalue to the dummy data frame
for (i in timepoint_list){
  popDRs <- dplyr::filter(get(i), empirical_qval < 0.05) %>% nrow() %>% as.numeric()
  popDR_count[[i]] <- popDRs
}

#remove dummy column, transpose axis, and add rownames to column
popDR_count <- dplyr::select(popDR_count, -dummy) %>% as.matrix() %>% t() %>% as.data.frame() %>% rownames_to_column("Sample") 

#extract timepoint and rename V1
popDR_count <- mutate(popDR_count, Timepoint = str_extract(Sample, "(?<=Timepoint)T[^:]+")) %>% rename("popDR_count" = V1)

#relevel timepoint factor
popDR_count$Timepoint <- factor(popDR_count$Timepoint, levels = c("T1","T2","T3","T4","T5","T6","T7","T8","T9","T10","T12","T14","T16","T18","T24","T30","T36","T42","T48"))
```

Plot
```{r}
popdr_plot <- ggplot(popDR_count, aes(x = Timepoint, y = popDR_count, group = 1)) +
  geom_line(color = "mediumpurple") +
  geom_text(aes(label = popDR_count), vjust = -0.5) + 
  geom_point(size = 3, color = "mediumpurple") +
  labs(y = "Number of popDRs", x  ="Timepoint",
       title = "Number of popDRs per timepoint", 
       subtitle = "Thresholds: qval < 0.05") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold"), plot.subtitle = element_text(hjust = 0.5, face = "italic")) +
  coord_cartesian(ylim = c(0, 8000))
popdr_plot
```

### What pathways are the popDR genes enriched for?

Load hallmark pathways
```{r}
#load Human Molecular Signatures Database (msigdb) pathways
hallmark_pathways <- fgsea::gmtPathways("../../common-use-files/msigdb_pathways/h.all.v2023.2.Hs.symbols.gmt")
```

Loop through the timepoints with over 100 popDRs.
```{r}
#setup list to iterate through
timepoint_list <- list()
for (t in c("T4","T5","T30","T36","T42","T48")) {
  time <- paste0("Timepoint", t, ":African_admixture:InfectionMtb")
  timepoint_list <- append(timepoint_list, time)
}

#make dummy list
gsea_list <- list()

for (i in timepoint_list) {
  #print element
  print(i)
  #assign to res
  res <- get(i)
  #rank by t stat
  rank <- dplyr::arrange(res, t_statistic)
  #isolate symbol from gene_id - run distinct to make sure there are no duplicates
  rank <- mutate(rank, symbol = str_split_i(gene_id, pattern = "_", i = 2)) %>% distinct()
  #make rank vector
  rank_vec <- rank$t_statistic
  names(rank_vec) <- rank$symbol
  #run fgsea
  fgseaRes <- fgsea::fgsea(hallmark_pathways, rank_vec, nproc=1)
  #append to list
  gsea_list[[i]] <- fgseaRes
  #remove vars
  rm(res, rank, rank_vec, fgseaRes)
}
```

Select pathways with padj < 0.001 and a NES > +/- 1.5 in at least one timepoint. 
```{r}
select_top_pathways <- function(res) {
  sig <- as.data.frame(res) %>% dplyr::filter(padj < 0.001)
  top <- dplyr::filter(sig, NES > 2)
  bottom <- dplyr::filter(sig, NES < -2)
  return(dplyr::bind_rows(top, bottom))
}

gsea_sig_list <- lapply(gsea_list, FUN = select_top_pathways)

sig_pathways <- c(gsea_sig_list$`TimepointT4:African_admixture:InfectionMtb`$pathway, 
                  gsea_sig_list$`TimepointT5:African_admixture:InfectionMtb`$pathway, 
                  gsea_sig_list$`TimepointT30:African_admixture:InfectionMtb`$pathway, 
                  gsea_sig_list$`TimepointT36:African_admixture:InfectionMtb`$pathway, 
                  gsea_sig_list$`TimepointT42:African_admixture:InfectionMtb`$pathway, 
                  gsea_sig_list$`TimepointT48:African_admixture:InfectionMtb`$pathway) %>% unique() %>% as.data.frame()
colnames(sig_pathways) <- "pathway"
```

Now that we have top pathways, extract NES from each timepoint and put into a dataframe
```{r}
for (i in timepoint_list) {
  #extract timepoint
  res <- gsea_list[[i]] %>% as.data.frame()
  #intersect with sig pathways
  res <- inner_join(res, sig_pathways, by = "pathway")
  #parse if fiirst time to add pathway
  if (i == "TimepointT4:African_admixture:InfectionMtb") {
    gsea_sig_df <- dplyr::transmute(res, pathway, !!i := NES)
  }
  if (i != "TimepointT4:African_admixture:InfectionMtb") {
  #add to a new column of df
    res <- dplyr::transmute(res, pathway, !!i := NES)
    gsea_sig_df <- left_join(gsea_sig_df, res, by = "pathway")
  }
  rm(res)
}
```

Pivot longer
```{r}
#order by mean NES
gsea_sig_df <- mutate(gsea_sig_df, mean_NES = rowMeans(across(-1))) %>%
  arrange((mean_NES)) %>%
  dplyr::select(1, everything()) %>% dplyr::select(-mean_NES)

#remove "hallmark_" from each pathway
gsea_sig_df <- mutate(gsea_sig_df, pathway = str_remove(pathway, "HALLMARK_"))

gsea_sig_df$pathway <- factor(gsea_sig_df$pathway, levels = rev(c(gsea_sig_df$pathway)))

gsea_sig_df <- pivot_longer(gsea_sig_df, cols = 2:last_col(), names_to = "Sample", values_to = "NES")

gsea_sig_df <- mutate(gsea_sig_df, Timepoint = str_extract(Sample, "(?<=Timepoint)T[^:]+"))

gsea_sig_df$Timepoint <- factor(gsea_sig_df$Timepoint, levels = c("T4","T5","T30","T36","T42","T48"))
```

Plot results as a heatmap with x=time, y=pathway, color=NES
```{r, fig.width=7.5, fig.height=8}
library(RColorBrewer)
my_palette <- colorRampPalette(rev(brewer.pal(11, "PuOr")))(50)

gsea_heatmap <- ggplot(gsea_sig_df, aes(x = Timepoint, y = pathway, fill = NES)) +
  geom_tile(color = "white", lwd = 0.5) +
  scale_fill_gradientn(colors = my_palette) +
  labs(title = "",
       subtitle = "*only pathways with padj < 0.001 and a NES > +/- 2\nin at least one timepoint are shown",
       x = "Time Post-Infection (hrs)", 
       y = "Hallmark Pathways") +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5, face = "bold", size = 12), 
        plot.subtitle = element_text(hjust = 0.5, face = "italic"), 
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  coord_equal()

gsea_heatmap
```


#### What pathways are specific to these clusters?
Pathway enrichment analysis. Use the expressed genes as the universe to identify unique pathways among each. 

```{r}
suppressPackageStartupMessages(library(clusterProfiler))
suppressPackageStartupMessages(library(ReactomePA))
suppressPackageStartupMessages(library(org.Hs.eg.db))
```

Up genes
```{r}
#empty list 
up_genes <- list()
#get genes from each relevant timepoint
for (i in timepoint_list){
  #filter genes by qval
  tmp1 <- dplyr::filter(get(i), empirical_qval < 0.05 & coefficients > 0) %>% mutate(ensembl = str_split(gene_id, "_", simplify = T)[,1])
  
  #get entrezid
  tmp2 <- bitr(tmp1$ensembl, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")
  
  #remove ensembl
  tmp2 <- tmp2$ENTREZID
    
  #append to list
  name_i <- paste0(i) %>% str_extract("(?<=Timepoint)T[^:]+")
  up_genes[[name_i]] <- tmp2
  rm(tmp1, tmp2)
}
```

down genes
```{r}
#empty list 
down_genes <- list()
#get genes from each relevant timepoint
for (i in timepoint_list){
  #filter genes by qval
  tmp1 <- dplyr::filter(get(i), empirical_qval < 0.05 & coefficients < 0) %>% mutate(ensembl = str_split(gene_id, "_", simplify = T)[,1])
  
  #get entrezid
  tmp2 <- bitr(tmp1$ensembl, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")
  
  #remove ensembl
  tmp2 <- tmp2$ENTREZID
    
  #append to list
  name_i <- paste0(i) %>% str_extract("(?<=Timepoint)T[^:]+")
  down_genes[[name_i]] <- tmp2
  rm(tmp1, tmp2)
}
```

Get expressed genes
```{r}
#Read in genes
expr_genes <- readRDS('data/filtered_counts.rds') %>% row.names() %>% as.data.frame()
colnames(expr_genes) <- "gene_id"

#get ensembl
expr_genes <- mutate(expr_genes, ensembl = str_split(gene_id, "_", simplify = T)[,1])

#get entrez
expr_ids <- bitr(expr_genes$ensembl, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")

#assign
universe <- expr_genes$ENTREZID
```

Run analysis for up and down 
```{r}
up_reactome <- clusterProfiler::compareCluster(geneClusters = up_genes, fun = enrichPathway, universe = universe)
up_kegg <- clusterProfiler::compareCluster(geneClusters = up_genes, fun = enrichKEGG, universe = universe)
up_bp <- clusterProfiler::compareCluster(geneClusters = up_genes, fun = enrichGO, universe = universe, OrgDb = org.Hs.eg.db, ont = "BP", keyType = "ENTREZID")

down_reactome <- clusterProfiler::compareCluster(geneClusters = down_genes, fun = enrichPathway, universe = universe)
down_kegg <- clusterProfiler::compareCluster(geneClusters = down_genes, fun = enrichKEGG, universe = universe)
down_bp <- clusterProfiler::compareCluster(geneClusters = down_genes, fun = enrichGO, universe = universe, OrgDb = org.Hs.eg.db, ont = "BP", keyType = "ENTREZID")
```

Plot
```{r, fig.height=6, fig.width=8}
plot_up_reactome <- dotplot(up_reactome, showCategory = 2) + ggtitle("Reactome Pathway Enrichment for AA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_up_kegg <- dotplot(up_kegg, showCategory = 4) + ggtitle("KEGG Pathway Enrichment for AA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_up_bp <- dotplot(up_bp, showCategory = 2) + ggtitle('Gene Onology Enrichment for AA-associated PopDRs', subtitle = "Biological Process Terms" ) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))

plot_down_reactome <- dotplot(down_reactome, showCategory = 2) + ggtitle("Reactome Pathway Enrichment for EA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_down_kegg <- dotplot(down_kegg, showCategory = 4) + ggtitle("KEGG Pathway Enrichment for EA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_down_bp <- dotplot(down_bp, showCategory = 2) + ggtitle('Gene Onology Enrichment for EA-associated PopDRs', subtitle = "Biological Process Terms" ) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))


plot_up_reactome
plot_up_kegg
plot_up_bp

plot_down_reactome
plot_down_kegg
plot_down_bp
```

It is not as clean as I would like. I will now repeat but use the whole protein coding genome as the universe. 


```{r}
#get pc genes
pc_gene_ids <- plyranges::filter(gtf, type == 'gene', gene_biotype == 'protein_coding') %>% as.data.frame() %>% dplyr::select(gene_id, gene_name)

#get entrez_ids for pc_genes
pc_ids <- bitr(pc_gene_ids$gene_id, fromType = "ENSEMBL", toType = "ENTREZID", OrgDb = "org.Hs.eg.db")
```
```{r}
#set universe to all protein coding genes
universe <- pc_ids$ENTREZID
```

```{r}
up_reactome <- clusterProfiler::compareCluster(geneClusters = up_genes, fun = enrichPathway, universe = universe)
up_kegg <- clusterProfiler::compareCluster(geneClusters = up_genes, fun = enrichKEGG, universe = universe)
up_bp <- clusterProfiler::compareCluster(geneClusters = up_genes, fun = enrichGO, universe = universe, OrgDb = org.Hs.eg.db, ont = "BP", keyType = "ENTREZID")

down_reactome <- clusterProfiler::compareCluster(geneClusters = down_genes, fun = enrichPathway, universe = universe)
down_kegg <- clusterProfiler::compareCluster(geneClusters = down_genes, fun = enrichKEGG, universe = universe)
down_bp <- clusterProfiler::compareCluster(geneClusters = down_genes, fun = enrichGO, universe = universe, OrgDb = org.Hs.eg.db, ont = "BP", keyType = "ENTREZID")
```

Plot
```{r, fig.height=6, fig.width=8}
plot_up_reactome <- dotplot(up_reactome, showCategory = 4) + ggtitle("Reactome Pathway Enrichment for AA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_up_kegg <- dotplot(up_kegg, showCategory = 4) + ggtitle("KEGG Pathway Enrichment for AA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_up_bp <- dotplot(up_bp, showCategory = 4) + ggtitle('Gene Onology Enrichment for AA-associated PopDRs', subtitle = "Biological Process Terms" ) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))

plot_down_reactome <- dotplot(down_reactome, showCategory = 4) + ggtitle("Reactome Pathway Enrichment for EA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_down_kegg <- dotplot(down_kegg, showCategory = 4) + ggtitle("KEGG Pathway Enrichment for EA-associated PopDRs")+
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))
plot_down_bp <- dotplot(down_bp, showCategory = 4) + ggtitle('Gene Onology Enrichment for EA-associated PopDRs', subtitle = "Biological Process Terms" ) +
  scale_y_discrete(labels = function(x) str_wrap(x, width = 50))


plot_up_reactome
plot_up_kegg
plot_up_bp

plot_down_reactome
plot_down_kegg
plot_down_bp
```

Okay last thing. Lets look at HLADR over time. 

```{r}
HLADR <- v_popDR$E["ENSG00000143226_FCGR2A", ] %>% as.data.frame() %>% rownames_to_column("sample") %>%
  left_join(rownames_to_column(sample_info, var = "sample"), by = "sample")

colnames(HLADR)[2] <- "voom_cts"

HLADR$Timepoint <- factor(HLADR$Timepoint, levels = c("T1", "T2","T3","T4","T5","T6","T7","T8","T9","T10","T12","T14","T16","T18","T24","T30","T36","T42","T48"))

ggplot(HLADR, aes(x = Timepoint, y = voom_cts, color = Infection, fill = self_reported_ethnicity)) +
  geom_boxplot(alpha = 0.5) +
  scale_fill_manual(values = c("darkorange", "darkblue")) +
  scale_color_manual(values = c("grey", "black")) +
  ylab("Normalized Read Count") +
  ggtitle("Fc Gamma Receptor IIa (FCGR2A)") +
  theme_bw()
```

```{r}
HLADR <- v_popDR$E["ENSG00000204287_HLA-DRA", ] %>% as.data.frame() %>% rownames_to_column("sample") %>%
  left_join(rownames_to_column(sample_info, var = "sample"), by = "sample")

colnames(HLADR)[2] <- "voom_cts"

HLADR$Timepoint <- factor(HLADR$Timepoint, levels = c("T1", "T2","T3","T4","T5","T6","T7","T8","T9","T10","T12","T14","T16","T18","T24","T30","T36","T42","T48"))

ggplot(HLADR, aes(x = Timepoint, y = voom_cts, color = Infection, fill = self_reported_ethnicity)) +
  geom_boxplot(alpha = 0.5) +
  scale_fill_manual(values = c("darkorange", "darkblue")) +
  scale_color_manual(values = c("grey", "black")) +
  ylab("Normalized Read Count") +
  ggtitle("MHC Class II, DR Alpha (HLA-DRA)") +
  theme_bw()
```
